Security considerations editorial (#229)
mlswg/mls-architecture: draft-ietf-mls-architecture.md

The MLS protocol provides per-sender chains of symmetric authenticated
encryption with additional data (AEAD) {{!RFC5116}} keys that are
generated from Group Secrets. Specifically, each epoch establishes
a per-sender "Ratchet Secret", which is then used to generate an

> AEAD key, which is used to protect MLS Plaintext messages.

Author: I don't think that this is gramatically correct. You don't generate the key to protect the messages.

> AEAD key, which is used to protect MLS Plaintext messages.

Reviewer: Please discard : )

Each time a message is sent, the Ratchet Secret is used

> to create a new Ratchet Secret and a new corresponding AEAD key.

Author: Well, I think we need to say something in order to explain the  security properties. How about

"A new Ratchet Secret is used to generate the AEAD keys for each new message"

> to create a new Ratchet Secret and a new corresponding AEAD key.

Reviewer: That works, yes !

> **RECOMMENDATION:** If the threat model of the system is against an adversary
> which can access the messages on the device without even needing to attack

> > MLS, the application should delete plaintext messages.

Author: OK. I can live with keeping this.

#### Authentication compromise: Ghost users and impersonations

One important property of MLS is that all Members know which other members are in
the group at all times. If all Members of the group and the Authentication
Service are honest, no parties other than the members of the current group can
read and write messages protected by the protocol for that Group.

This guarentee applies to the the cryptographic identities of the members,

> but these identities need to be bound to people in the physical world.

Reviewer: MLS being used in machine to machine cases, they don't.

> but these identities need to be bound to people in the physical world.

Author: OK.

Details about how to verify the identity of a client depend on the MLS
Credential type used. For example, cryptographic verification of credentials can
be largely performed autonomously (e.g., without user interaction) by
the clients themselves for the `x509` Credential
type.

In contrast, when MLS clients use the `basic` Credential type, then some
other mechanism must be used to verify identities. For instance the Authentication
Service could operate some sort of directory server to provide keys,

> or users could potentially verify keys via some out-of-band mechanism.

Reviewer: ```suggestion
or users could verify keys via an out-of-band mechanism.
```

other mechanism must be used to verify identities. For instance the Authentication
Service could operate some sort of directory server to provide keys,
or users could verify keys via an out-of-band mechanism.

> **RECOMMENDATION:** Select the strongest MLS Credential type available among
> the target members of an MLS group.

If the AS is compromised, it could validate a (or generate a new)
signature keypair for an attacker. Because a user can have many MLS
clients running the MLS protocol, it possibly has many signature
keypairs for multiple devices. These attacks could be very difficult
to detect, especially in large groups where the UI might not reflect
all the changes back to the users. If the application participates in
a key transparency mechanism in which it is possible to determine
every key for a given user, then this then this would allow for the
detection of surreptitiously created false credentials.

> **RECOMMENDATION:** Make sure that MLS clients reflect all the membership
> changes to the users as they happen. If a choice has to be made because the
> number of notifications is too high, the client should provide a log of

> > state of the device so that the user can examine it.

Author: Hmmm..... ISTM that there are two potential things this might be talking about:

1. Changes to the set of clients in a group.
2. Changes to the set of clients for a user, whether in a group or not.

I assumed that this text was about the former because of the text "This is especially the case in large groups where the UI might not reflect all the changes back to the users." which is clearly about groups, not the set of clients for a user.

I agree that it would be fine to talk about a transparency log for user credentials, but then this needs a substantial rewrite.

to prove their identity to each other. This can be done for instance using a QR
code that can be scanned by the other parties.

Another way to improve the security for the users is to provide a transparency
mechanism which allows each user to check if credentials used in groups have
been published in the transparency log. Another benefit of this mechanism is for
revocation. The users of a group could check for revoked keys (in case of
compromise detection) using a mechanism such as CRLite or some more advanced

> privacy preserving technology.

Reviewer 2: Add back some mention of revocation. Maybe:

"When the credential type support revocation, the users of a group could check for revoked keys, for example for `x509` credentials, using a mechanism such as CRLite."

(I don't particularly care if we mention CRLite or some other mechanism, but I think we should say something about revocation.)
